lib/corefoundation/exceptions.rb

```diff
@@ -0,0 +1,16 @@
+module CF
+  module Exceptions
+    # Raised when the preference value couldn't be found for a domain.
+    class PreferenceDoesNotExist < RuntimeError
+      def initialize(key, domain, hostname)
```

Comment:
I guess the first arg is meant for `msg`, let's keep it like that. Having a default/generic `msg` would help too.
Also, let's make the other args optional.

Reply:
The whole idea behind having a restricted error class is they are used in only specific places and are "idiomatic". They serve a specific purpose - to raise that error. Think of it as sort of helpers. For another example, please take a look at this https://github.com/chef/chef/blob/master/lib/chef/exceptions.rb

Reply:
If you are looking for helper functionality, extract it out as a `utility method` or something. Having constructor overrides unless you are assigning custom properties to the error object itself is not of much use here.

Reply:
It is not meant to be a helper. A helper would have been generic. This is meant to be specific and give an accurate description of the kind of error it is. But I agree with the `args` part though, we could probably make that variable. I guess I confused you in my previous comment by asking you to think of it as a helper.

Reply:
I agree it didn't make much sense in the first look. But once I looked at how these are used in chef, these seemed like a good way to implement it rather than having a helper that does too much as the code grows.

Reply:
Try something like this
```ruby
def initialize(msg = 'some default msg', ...optional args)
```

Reply:
I'm afraid you still don't see the point I'm trying to make here. If I wanted the error message to be modified, I wouldn't need this since I could just write `raise "error message"` and that would have worked. `PreferenceDoesNotExist` is meant to cater to a very specific error so it doesn't need to take message as input. The benefit of using custom exceptions like this would be that now we have a category that multiple errors could potentially belong to. At that point too, the logic for building the message would be internal to this subclass and shouldn't require anything else apart from the context.

Reply:
Let me clarify what I want to support here,
```ruby
raise PreferenceDoesNotExist # hence the default msg, this is optional/good to have
raise PreferenceDoesNotExist, 'some error msg',
raise PreferenceDoesNotExist, 'some error msg', ...optional args
```

You can now use this error class for any pref search, irrespective of username/hostname